features/net/network-socket/nsapi_dns.h

```diff
@@ -0,0 +1,103 @@
+/* nsapi_dns.h
+ * Original work Copyright (c) 2013 Henry Leinen (henry[dot]leinen [at] online [dot] de)
+ * Modified work Copyright (c) 2015 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef NSAPI_DNS_H
+#define NSAPI_DNS_H
+
+#include "nsapi_types.h"
+#include "network-socket/NetworkStack.h"
+
+
+/** Query a domain name server for an IP address of a given hostname
+ *
+ *  @param stack    Network stack as target for DNS query
+ *  @param addr     Destination for the host address
+ *  @param host     Hostname to resolve
+ *  @param version  IP version to resolve (defaults to NSAPI_IPv4)
+ *  @return         0 on success, negative error code on failure
+ *                  NSAPI_ERROR_DNS_FAILURE indicates the host could not be found
+ */
+NSAPI_C_LINKAGE
+int nsapi_dns_query(nsapi_stack_t *stack, nsapi_addr_t *addr,
+        const char *host, nsapi_version_t version);
+
+#ifdef __cplusplus
```

Comment:
The original goal was to avoid duplicated function definitions.

If we don't care about duplicated code, the header file could be restructured into C and C++ sections, that would reduce the noise in the file. I would vote against spliting the header file since that would just move the noise to the files in the directory.

C and C++ actually play together quite well as long as use `extern "C"` carefully (this is what NSAPI_C_LINKAGE abstracts away) and C++ types are hidden behind opaque pointers. In this case the language juggling is needed to workaround the fact that we have no way to cast the `NetworkStack` class into a C style `nsapi_stack_t` and that nsapi_dns is intended for use in the network stack layer, which has shown to be generally C.